### ISSUE TITLE
[Feature] Make submodules’ dtype be configurable

### ISSUE THREAD

**Issue by aolemila** [sgl-project/sglang-jax#661]
### **Checklist**

- [ ] If the issue you raised is not a feature but a question, please raise a discussion at https://github.com/sgl-project/sgl-jax/discussions/new/choose Otherwise, it will be closed.
- [ ] Please use English, otherwise it will be closed.

### **Motivation**

Control the precision scope when using higher precision, like float32.

* softmax
* proj in attn and mlp

### **Related resources**

**Comment by iwknow** [comment]
Can you please provide more details about this feature? maybe i can work on it.

**Comment by iwknow** [comment]
let my summarize my understanding, @aolemila please let me know if it is correct.

Basically, this feature request is about allowing more precise control on the dtype on each submodule of a model. Use python/sgl_jax/srt/models/llama.py as an example, While `LlamaDecoderLayer` currently has a dtype parameter, the granularity of that configuration is limited: a single data type is applied to every component within a layer. In `LlamaAttention`, the `q_proj`, `k_proj`, `v_proj`, and `o_proj` are all initialized with the same dtype provided to the parent layer. You cannot currently specify, for example, that the output projection (o_proj) should use `float32` for while the others remain in `bfloat16`.  so we'd like to introduce a `precision_config` or `dtype_config` so that the dtype for each of the components can be specified.

**Comment by iwknow** [comment]
i'd love to. i will provide my design and implementation plan shortly.

**Comment by aolemila** [comment]
Hi @iwknow , thanks for your design. Here are some suggestions.

1. Please refer to [contribution document](https://github.com/sgl-project/sglang-jax/blob/main/docs/developer_guide/contribution_guide.md). I should provide you early. Move the design into a google document and create a subissue to put your document link. After our discussion, it is better to create a pr to put your document content under the docs/feature/xxx.md. Then you can create another subissue to do the development. You can refer to this #639.

2. In the design document, you can refer to [this one](https://docs.google.com/document/d/1ei4LWKQsQoQB87LCNSfLCMPEbnxPWrewdUVlWalO3rk/edit?tab=t.0#heading=h.gtpewejqczi). It is better to include the user story, design, test plan.

3. Take the `lllama.py` as an example, `rotary_emb` is missing in `self_attn`. Tips: qwen2 and qwen3 use o_proj and qwen uses c_proj. `c_fc` does not exist in `qwen.py` or `llama.py`.

4. In addition to the above field config, could you give more details about how to modify the codes to let the config work in engine. The base principle is as much as not to influence the existing model codes.

**Comment by aolemila** [comment]
Hi, @iwknow , I have reviewed your document. Please take a look at comments.

**Comment by aolemila** [comment]
@JamesBrianD @wang2yn84 If you have time, please take a look at this feature.

**Comment by aolemila** [comment]
Hi @Iamleos ! If you have time, please take a look at softmax in `raged_paged_attention.py`. We do the softmax according flash attention v2. We prefer to use float32 as the dtype result of q @ k. And the s will be used to do the next logic including softmax. So I think we do not need to add extra logic which is similar to `native_backend.py` in `raged_paged_attention.py`. What about your opinion?

Code: `s = jnp.einsum("nd,md->nm", q, k, preferred_element_type=jnp.float32)`
